File: backend/core/src/modules/research-items/research-items.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { ProjectScopedCrudService } from '../../common/access/project-scoped-crud.service';
import { ResearchItem } from '../../entities';
import { AiService } from '../ai/ai.service';

@Injectable()
export class ResearchItemsService extends ProjectScopedCrudService<ResearchItem> {
  constructor(
    @InjectRepository(ResearchItem)
    repo: Repository<ResearchItem>,
    projectAccessService: ProjectAccessService,
    private readonly aiService: AiService,
  ) {
    super(repo, projectAccessService);
  }

  async generateForUser(
    user: { userId: string; role?: string },
    dto: { projectId?: string; query?: string; provider?: string; model?: string },
  ) {
    if (!dto.projectId) {
      throw new BadRequestException('projectId가 필요합니다.');
    }
    if (!dto.query) {
      throw new BadRequestException('query가 필요합니다.');
    }

    await this.projectAccessService.assertProjectAccess(user.userId, dto.projectId);

    const ai = await this.aiService.proxy(user, 'research', '/ai/research', {
      query: dto.query,
      provider: dto.provider,
      model: dto.model,
      projectId: dto.projectId,
    });

    const entity = this.repo.create({
      projectId: dto.projectId,
      query: dto.query,
      result: ai as any,
    } as any);
    return this.repo.save(entity as any);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/ResearchItemsService.java
package store.pjcloud.cowrite.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.http.HttpStatus;
import store.pjcloud.cowrite.core.common.ProjectScopedCrudService;
import store.pjcloud.cowrite.core.entity.ResearchItem;
import store.pjcloud.cowrite.core.repository.ResearchItemRepository;

@Service
public class ResearchItemsService extends ProjectScopedCrudService<ResearchItem> {
    private final ProjectAccessService projectAccessService;
    private final AiService aiService;
    private final ObjectMapper objectMapper;

    public ResearchItemsService(ResearchItemRepository repository,
                                ProjectAccessService projectAccessService,
                                ObjectMapper objectMapper,
                                AiService aiService) {
        super(repository, projectAccessService, objectMapper, ResearchItem.class);
        this.projectAccessService = projectAccessService;
        this.aiService = aiService;
        this.objectMapper = objectMapper;
    }

    public ResearchItem generateForUser(AiService.UserContext user, Map<String, Object> body) {
        Object projectIdObj = body.get("projectId");
        Object queryObj = body.get("query");
        if (!(projectIdObj instanceof String) || ((String) projectIdObj).isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "projectId가 필요합니다.");
        }
        if (!(queryObj instanceof String) || ((String) queryObj).isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "query가 필요합니다.");
        }
        UUID projectId = UUID.fromString((String) projectIdObj);
        projectAccessService.assertProjectAccess(user.userId(), projectId);

        Object ai = aiService.proxy(user, "research", "/ai/research", body);
        ResearchItem item = new ResearchItem();
        item.setProjectId(projectId);
        item.setQuery((String) queryObj);
        item.setResult(objectMapper.convertValue(ai, Map.class));
        return repository().save(item);
    }
}
